### PR TITLE
Navigation: Fix Home logo always going to `/login`

### LINKED ISSUE #57048: click on the top left grafana icon navigate to the login page for unauthenticated users
**What happened**:
An unauthenticated user go to [play.grafana.org](https://play.grafana.org/) and click on the top left grafana icon, then it navigates to the login page.
An authenticated user would juste return to the home page, like in the previous versions of grafana.

**What you expected to happen**:
I would expect authenticated and unauthenticated to return to the home page.

**How to reproduce it (as minimally and precisely as possible)**:
go to [play.grafana.org](https://play.grafana.org/) click on grafana icon

**Anything else we need to know?**:
no

**Environment**:
- Grafana version: 9.2.0
- Data source type & version: none
- OS Grafana is installed on: none
- User OS & Browser: all
- Grafana plugins: none
- Others:

### PR DESCRIPTION
**Pull request by ashharrison90**
**What is this feature?**

- exposes whether anonymous access is enabled or not in the `bootData`
- changes the home nav item to redirect to `/login` if not logged in and anonymous access isn't enabled
- changes the home logo to redirect to `/login` if not logged in and anonymous access isn't enabled
- stops command palette calling search for recent dashboards if not logged in
- stops command palette searching for dashboards when typed in if not logged in and anonymous access isn't enabled

**Why do we need this feature?**

- better behaviour when anonymous access is enabled (e.g. play)
- better behaviour when viewing a snapshot and anonymous access isn't enabled (no access denied modal due to command palette making background calls)

**Who is this feature for?**

- everyone

**Which issue(s) does this PR fix?**:

Fixes #57048 

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/core/components/NavBar/NavBar.tsx [hunk 1]
```diff
@@ -51,11 +51,16 @@ export const NavBar = React.memo(() => {
     menuOpen
   );
 
+  let homeUrl = config.appSubUrl || '/';
+  if (!config.bootData.user.isSignedIn && !config.anonymousEnabled) {
+    homeUrl = locationUtil.getUrlForPartial(location, { forceLogin: 'true' });
```
Comment (torkelo) [inline public/app/core/components/NavBar/NavBar.tsx:56]: when is this ever true? can't the logo always go to `config.appSubUrl || '/'` , what is this scenario for?

**Comment by joshhunt** [comment]
(would be better if the command palette stuff was in a separate PR..... 🥺 )

> stops command palette searching for dashboards when typed in if not logged in and anonymous access isn't enabled

Shouldn't this already be handled by the [checks for not rendering CommandPalette](https://github.com/grafana/grafana/blob/main/public/app/AppWrapper.tsx#L93-L96)? Or is there a scenario I'm not thinking of?

File: public/app/core/components/NavBar/NavBar.tsx (hunk 1 above)
Comment (ashharrison90) [reply public/app/core/components/NavBar/NavBar.tsx:56]: it's true for snapshots and maybe(?) public dashboards (haven't checked this one yet)

**Comment by ashharrison90** [comment]
> (would be better if the command palette stuff was in a separate PR..... 🥺 )

yeah i wanted to put it in a separate PR, but both changes relied on the backend changes to expose `anonymousEnabled` so kept them together in the end else they'd end up conflicting/relying on each other anyway

> Shouldn't this already be handled by the [checks for not rendering CommandPalette](https://github.com/grafana/grafana/blob/main/public/app/AppWrapper.tsx#L93-L96)? Or is there a scenario I'm not thinking of?

it definitely doesn't handle it right now. that just prevents showing the command palette on the login page. we could change that logic to disable the command palette if not logged in and anonymous access isn't enabled? i think that would be ok :)

File: public/app/core/components/NavBar/NavBar.tsx (hunk 1 above)
Comment (joshhunt) [reply public/app/core/components/NavBar/NavBar.tsx:56]: Old navigation (before megamenu iirc) used to have this behaviour - we regressed this quite a bit ago.

**Review by mckn — COMMENTED** [review]
Looks good to me!

Would it work to expose the `homeUrl` in the boot data instead of exposing `anonymousEnabled`? It would have been nice to have the logic of what value the homeUrl should be in one place.

**Comment by ashharrison90** [comment]
@mckn we kinda do already. it's exposed as the `url` on the home node in `navTree`. **but** we need to expose `anonymousEnabled` anyway to control whether to allow dashboard searching [here](https://github.com/grafana/grafana/pull/62658/files#diff-8186f5baf18f99985dc3647e712e050e3ccf4dd816ac951c106038344ce10beaR55).

i have tried using just this url in `NavBar`/`TopSearchBar`, but using the full logic has 2 benefits:
  - redirects you back to the page you were on originally when you clicked home
  - correctly sets the chromeless state on the login page

**Comment by mckn** [comment]
> @mckn we kinda do already. it's exposed as the `url` on the home node in `navTree`. **but** we need to expose `anonymousEnabled` anyway to control whether to allow dashboard searching [here](https://github.com/grafana/grafana/pull/62658/files#diff-8186f5baf18f99985dc3647e712e050e3ccf4dd816ac951c106038344ce10beaR55).
> 
> i have tried using just this url in `NavBar`/`TopSearchBar`, but using the full logic has 2 benefits:
> 
> * redirects you back to the page you were on originally when you clicked home
> * correctly sets the chromeless state on the login page

Ah, ok ok! Got it. Then I would say this is good to go from my side.

**Comment by monotek** [comment]
Any chance to see thsi merged soonish?

**Comment by ashharrison90** [comment]
@monotek working on it 👍

File: public/app/features/commandPalette/actions/dashboardActions.ts [hunk 2]
```diff
@@ -15,6 +17,10 @@ const MAX_RECENT_DASHBOARDS = 5;
 const debouncedSearch = debounce(getSearchResultActions, 200);
 
 export async function getRecentDashboardActions(): Promise<CommandPaletteAction[]> {
+  if (!contextSrv.user.isSignedIn) {
```
Comment (tskarhed) [inline public/app/features/commandPalette/actions/dashboardActions.ts:20]: Could we maybe add a test for this one?

**Review by tskarhed — APPROVED** [review]
Nice job with the tests! :+1:

**Comment by DanCech** [comment]
This PR has triggered 2 CodeQL warnings for cross-site scripting because we're now using the user-provided location.pathname to build the url:

https://github.com/grafana/grafana/security/code-scanning?query=pr%3A65069+tool%3ACodeQL+is%3Aopen

**Comment by ashharrison90** [comment]
@DanCech thanks for pointing that out. fix here: https://github.com/grafana/grafana/pull/65293

**Pull request merged**